File: rxhttp-compiler/src/main/java/com/rxhttp/compiler/AnnotationProcessor.java
package com.rxhttp.compiler;

import com.rxhttp.compiler.exception.ProcessingException;

import java.io.IOException;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Filer;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import javax.tools.Diagnostic;

import rxhttp.wrapper.annotation.Converter;
import rxhttp.wrapper.annotation.DefaultDomain;
import rxhttp.wrapper.annotation.Domain;
import rxhttp.wrapper.annotation.Param;
import rxhttp.wrapper.annotation.Parser;

/**
 * User: ljx
 * Date: 2019/3/21
 * Time: 20:36
 */
public class AnnotationProcessor extends AbstractProcessor {

    private Types typeUtils;
    private Messager messager;
    private Filer filer;
    private Elements elementUtils;
    private boolean processed;
    private String platform;

    @Override
    public synchronized void init(ProcessingEnvironment processingEnvironment) {
        super.init(processingEnvironment);
        typeUtils = processingEnvironment.getTypeUtils();
        messager = processingEnvironment.getMessager();
        filer = processingEnvironment.getFiler();
        elementUtils = processingEnvironment.getElementUtils();

        Map<String, String> map = processingEnvironment.getOptions();
        platform = map.get("platform");
        if (platform == null) {
            platform = "Android";
        }
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        Set<String> annotations = new LinkedHashSet<>();
        annotations.add(Param.class.getCanonicalName());
        annotations.add(Parser.class.getCanonicalName());
        annotations.add(Converter.class.getCanonicalName());
        annotations.add(Domain.class.getCanonicalName());
        annotations.add(DefaultDomain.class.getCanonicalName());
        annotations.add(Override.class.getCanonicalName());
        return annotations;
    }

    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }


    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
//        messager.printMessage(Kind.WARNING, "process start annotations" + annotations + " this=" + this);
        if (annotations.isEmpty() || processed) return true;
        try {
            RxHttpGenerator rxHttpGenerator = new RxHttpGenerator();

            ParamsAnnotatedClass paramsAnnotatedClass = new ParamsAnnotatedClass();
            for (Element element : roundEnv.getElementsAnnotatedWith(Param.class)) {
                TypeElement typeElement = (TypeElement) element;
                checkParamsValidClass(typeElement);
                paramsAnnotatedClass.add(typeElement);
            }

            ParserAnnotatedClass parserAnnotatedClass = new ParserAnnotatedClass();
            for (Element element : roundEnv.getElementsAnnotatedWith(Parser.class)) {
                TypeElement typeElement = (TypeElement) element;
                checkParserValidClass(typeElement);
                parserAnnotatedClass.add(typeElement);
            }

            ConverterAnnotatedClass converterAnnotatedClass = new ConverterAnnotatedClass();
            for (Element annotatedElement : roundEnv.getElementsAnnotatedWith(Converter.class)) {
                VariableElement variableElement = (VariableElement) annotatedElement;
                checkConverterValidClass(variableElement);
                converterAnnotatedClass.add(variableElement);
            }

            DomainAnnotatedClass domainAnnotatedClass = new DomainAnnotatedClass();
            for (Element annotatedElement : roundEnv.getElementsAnnotatedWith(Domain.class)) {
                VariableElement variableElement = (VariableElement) annotatedElement;
                checkVariableValidClass(variableElement);
                domainAnnotatedClass.add(variableElement);
            }

            Set<? extends Element> elementSet = roundEnv.getElementsAnnotatedWith(DefaultDomain.class);
            if (elementSet.size() > 1)
                throw new ProcessingException(elementSet.iterator().next(), "@DefaultDomain annotations can only be used once");
            else if (elementSet.iterator().hasNext()) {
                VariableElement variableElement = (VariableElement) elementSet.iterator().next();
                checkVariableValidClass(variableElement);
                rxHttpGenerator.setAnnotatedClass(variableElement);
            }
            rxHttpGenerator.setAnnotatedClass(paramsAnnotatedClass);
            rxHttpGenerator.setAnnotatedClass(parserAnnotatedClass);
            rxHttpGenerator.setAnnotatedClass(converterAnnotatedClass);
            rxHttpGenerator.setAnnotatedClass(domainAnnotatedClass);

            // Generate code
            rxHttpGenerator.generateCode(elementUtils, filer, platform);
            processed = true;
        } catch (ProcessingException e) {
            error(e.getElement(), e.getMessage());
        } catch (IOException e) {
            e.printStackTrace();
        }

        return true;
    }

    private void checkParamsValidClass(TypeElement element) throws ProcessingException {
        if (!element.getModifiers().contains(Modifier.PUBLIC)) {
            throw new ProcessingException(element,
                "The class %s is not public",
                Param.class.getSimpleName());
        }
        if (element.getModifiers().contains(Modifier.ABSTRACT)) {
            throw new ProcessingException(element,
                "The class %s is abstract. You can't annotate abstract classes with @%",
                element.getSimpleName().toString(), Param.class.getSimpleName());
        }

        TypeElement currentClass = element;
        while (true) {
            List<? extends TypeMirror> interfaces = currentClass.getInterfaces();
            for (TypeMirror typeMirror : interfaces) {
                if (!typeMirror.toString().equals("rxhttp.wrapper.param.Param<P>")) continue;
                return;
            }
            TypeMirror superClassType = currentClass.getSuperclass();

            if (superClassType.getKind() == TypeKind.NONE) {
                throw new ProcessingException(element,
                    "The class %s annotated with @%s must inherit from %s",
                    element.getQualifiedName().toString(), Param.class.getSimpleName(),
                    "rxhttp.wrapper.param.Param");
            }
            currentClass = (TypeElement) typeUtils.asElement(superClassType);
        }
    }

    private void checkParserValidClass(TypeElement element) throws ProcessingException {
        if (!element.getModifiers().contains(Modifier.PUBLIC)) {
            throw new ProcessingException(element,
                "The class %s is not public",
                Parser.class.getSimpleName());
        }
        if (element.getModifiers().contains(Modifier.ABSTRACT)) {
            throw new ProcessingException(element,
                "The class %s is abstract. You can't annotate abstract classes with @%",
                element.getSimpleName().toString(), Parser.class.getSimpleName());
        }

        TypeElement currentClass = element;
        while (true) {
            List<? extends TypeMirror> interfaces = currentClass.getInterfaces();
            //遍历实现的接口有没有Parser接口
            for (TypeMirror typeMirror : interfaces) {
                if (typeMirror.toString().contains("rxhttp.wrapper.parse.Parser")) {
                    return;
                }
            }
            //未遍历到Parser，则找到父类继续，一直循环下去，直到最顶层的父类
            TypeMirror superClassType = currentClass.getSuperclass();
            if (superClassType.getKind() == TypeKind.NONE) {
                throw new ProcessingException(element,
                    "The class %s annotated with @%s must inherit from %s",
                    element.getQualifiedName().toString(), Parser.class.getSimpleName(),
                    "rxhttp.wrapper.parse.Parser<T>");
            }
            //TypeMirror转TypeElement
            currentClass = (TypeElement) typeUtils.asElement(superClassType);
        }
    }

    private void checkConverterValidClass(VariableElement element) throws ProcessingException {
        if (!element.getModifiers().contains(Modifier.PUBLIC)) {
            throw new ProcessingException(element,
                "The variable %s is not public",
                element.getSimpleName());
        }
        if (!element.getModifiers().contains(Modifier.STATIC)) {
            throw new ProcessingException(element,
                "The variable %s is not static",
                element.getSimpleName().toString());
        }
        TypeMirror classType = element.asType();
        if (!"rxhttp.wrapper.callback.IConverter".equals(classType.toString())) {
            while (true) {
                //TypeMirror转TypeElement
                TypeElement currentClass = (TypeElement) typeUtils.asElement(classType);
                //遍历实现的接口有没有IConverter接口
                for (TypeMirror mirror : currentClass.getInterfaces()) {
                    if (mirror.toString().equals("rxhttp.wrapper.callback.IConverter")) {
                        return;
                    }
                }
                //未遍历到IConverter，则找到父类继续，一直循环下去，直到最顶层的父类
                classType = currentClass.getSuperclass();
                if (classType.getKind() == TypeKind.NONE) {
                    throw new ProcessingException(element,
                        "The variable %s is not a IConverter",
                        element.getSimpleName().toString());
                }
            }
        }

    }

    private void checkVariableValidClass(VariableElement element) throws ProcessingException {
        if (!element.getModifiers().contains(Modifier.PUBLIC)) {
            throw new ProcessingException(element,
                "The variable %s is not public",
                element.getSimpleName());
        }
        if (!element.getModifiers().contains(Modifier.STATIC)) {
            throw new ProcessingException(element,
                "The variable %s is not static",
                element.getSimpleName().toString());
        }
    }

    private void error(Element e, String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.ERROR, String.format(msg, args), e);
    }
}

File: rxhttp-compiler/src/main/java/com/rxhttp/compiler/AnnotationProcessor.kt
package com.rxhttp.compiler

import com.rxhttp.compiler.ClassHelper.generatorStaticClass
import com.rxhttp.compiler.exception.ProcessingException
import net.ltgt.gradle.incap.IncrementalAnnotationProcessor
import net.ltgt.gradle.incap.IncrementalAnnotationProcessorType
import rxhttp.wrapper.annotation.*
import java.util.*
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.TypeElement
import javax.lang.model.element.VariableElement
import javax.lang.model.util.Elements
import javax.lang.model.util.Types
import javax.tools.Diagnostic


lateinit var rxHttpPackage: String  //RxHttp相关类的包名

/**
 * User: ljx
 * Date: 2019/3/21
 * Time: 20:36
 */
//@SupportedOptions(value = ["rxhttp_rxjava", "rxhttp_package", "rxhttp_incremental", "rxhttp_debug"])
@IncrementalAnnotationProcessor(IncrementalAnnotationProcessorType.DYNAMIC)
open class AnnotationProcessor : AbstractProcessor() {
    companion object {
        const val rxhttp_rxjava = "rxhttp_rxjava"
        const val rxhttp_package = "rxhttp_package"
        const val rxhttp_incremental = "rxhttp_incremental"
        const val rxhttp_debug = "rxhttp_debug"
    }

    private lateinit var types: Types
    private lateinit var messager: Messager
    private lateinit var filer: Filer
    private lateinit var elementUtils: Elements
    private var debug = false
    private var processed = false
    private var incremental = true

    @Synchronized
    override fun init(processingEnvironment: ProcessingEnvironment) {
        super.init(processingEnvironment)
        types = processingEnvironment.typeUtils
        messager = processingEnvironment.messager
        filer = processingEnvironment.filer
        elementUtils = processingEnvironment.elementUtils
        val map = processingEnvironment.options
        rxHttpPackage = map[rxhttp_package] ?: "rxhttp.wrapper.param"
        incremental = "false" != map[rxhttp_incremental]
        debug = "true" == map[rxhttp_debug]
        initRxJavaVersion(getRxJavaVersion(map))
    }

    override fun getSupportedAnnotationTypes(): Set<String> {
        return LinkedHashSet<String>().apply {
            add(Override::class.java.canonicalName)
            add(Param::class.java.canonicalName)
            add(Parser::class.java.canonicalName)
            add(Converter::class.java.canonicalName)
            add(Domain::class.java.canonicalName)
            add(DefaultDomain::class.java.canonicalName)
            add(OkClient::class.java.canonicalName)
        }
    }

    override fun getSupportedOptions(): MutableSet<String> {
        return mutableSetOf(
            rxhttp_rxjava,
            rxhttp_package,
            rxhttp_incremental,
            rxhttp_debug
        ).apply {
            if (incremental) {
                add("org.gradle.annotation.processing.aggregating")
            }
        }
    }

    open fun getRxJavaVersion(map: Map<String, String>) = map[rxhttp_rxjava]

    open fun isAndroidPlatform() = true

    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()

    override fun process(annotations: Set<TypeElement>, roundEnv: RoundEnvironment): Boolean {
        if (debug) {
            messager.printMessage(
                Diagnostic.Kind.WARNING,
                """
                    process isOver = ${roundEnv.processingOver()}     
                    processed = $processed                
                    rootElements.size = ${roundEnv.rootElements.size}
                    annotations = $annotations                     
                """.trimIndent()
            )
        }
        if (annotations.isEmpty() || processed) return true
        generatorStaticClass(filer, isAndroidPlatform())
        try {
            val rxHttpWrapper = RxHttpWrapper()

            val paramsVisitor = ParamsVisitor().apply {
                roundEnv.getElementsAnnotatedWith(Param::class.java).forEach {
                    val typeElement = it as TypeElement
                    add(typeElement, types)
                    rxHttpWrapper.add(typeElement)
                }
            }

            val parserVisitor = ParserVisitor().apply {
                roundEnv.getElementsAnnotatedWith(Parser::class.java).forEach {
                    val typeElement = it as TypeElement
                    add(typeElement, types)
                }
            }

            val converterVisitor = ConverterVisitor().apply {
                roundEnv.getElementsAnnotatedWith(Converter::class.java).forEach {
                    val variableElement = it as VariableElement
                    add(variableElement, types)
                    rxHttpWrapper.addConverter(variableElement)
                }
            }

            val okClientVisitor = OkClientVisitor().apply {
                roundEnv.getElementsAnnotatedWith(OkClient::class.java).forEach {
                    val variableElement = it as VariableElement
                    add(variableElement)
                    rxHttpWrapper.addOkClient(variableElement)
                }
            }

            val domainVisitor = DomainVisitor().apply {
                roundEnv.getElementsAnnotatedWith(Domain::class.java).forEach {
                    val variableElement = it as VariableElement
                    add(variableElement)
                    rxHttpWrapper.addDomain(variableElement)
                }
            }

            val defaultDomainVisitor = DefaultDomainVisitor().apply {
                set(roundEnv.getElementsAnnotatedWith(DefaultDomain::class.java))
            }

            //Generate RxHttp.java
            RxHttpGenerator().apply {
                this.paramsVisitor = paramsVisitor
                this.parserVisitor = parserVisitor
                this.converterVisitor = converterVisitor
                this.domainVisitor = domainVisitor
                this.defaultDomainVisitor = defaultDomainVisitor
                this.okClientVisitor = okClientVisitor
            }.generateCode(filer)

            // 生成 RxHttp 封装类
            rxHttpWrapper.generateRxWrapper(filer)
            processed = true
        } catch (e: ProcessingException) {
            error(e.element, e.message)
        } catch (e: Throwable) {
            e.printStackTrace()
            messager.printMessage(Diagnostic.Kind.ERROR, e.message)
        }
        return true
    }

    private fun error(e: Element, msg: String?, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.ERROR, String.format(msg ?: "", *args), e)
    }
}